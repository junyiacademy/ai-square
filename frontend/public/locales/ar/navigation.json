{
  "home": "الرئيسية",
  "dashboard": "لوحة التحكم",
  "relations": "العلاقات",
  "ksa": "إطار KSA",
  "assessment": "تقييم",
  "pbl": "التعلم بالمشاريع",
  "assessmentHistory": "سجل التقييمات",
  "history": "السجل",
  "signIn": "تسجيل الدخول",
  "signOut": "تسجيل الخروج",
  "userRole": {
    "student": "طالب",
    "teacher": "معلم",
    "admin": "مدير"
  },
  "darkMode": "التبديل إلى الوضع الداكن",
  "lightMode": "التبديل إلى الوضع الفاتح",
  "toggleTheme": "تبديل السمة",
  "historySubtitle": "عرض جميع سجلات أنشطة التعلم الخاصة بك",
  "totalActivities": "إجمالي الأنشطة",
  "filterAll": "الكل",
  "noHistory": "لا توجد سجلات تعلم حتى الآن",
  "startAssessment": "بدء التقييم",
  "startPBL": "بدء التعلم بالمشاريع",
  "startNewPBL": "بدء تعلم جديد بالمشاريع",
  "footer": {
    "product": "المنتج",
    "resources": "الموارد",
    "company": "الشركة",
    "about": "حول AI Square",
    "description": "AI Square هي منصة تعلم قائمة على Git لتعليم محو الأمية في الذكاء الاصطناعي، وتدعم 9 لغات حول العالم.",
    "journey": "رحلة المستخدم",
    "roadmap": "خارطة طريق المنتج",
    "feedback": "التعليقات",
    "rights": "جميع الحقوق محفوظة.",
    "privacy": "سياسة الخصوصية",
    "terms": "شروط الخدمة",
    "contact": "اتصل بنا",
    "languages": "اللغات",
    "domains": "مجالات الذكاء الاصطناعي",
    "competencies": "الكفاءات",
    "madeWith": "صنع بـ ❤️ لتعليم محو الأمية في الذكاء الاصطناعي"
  },
  "discovery": "الاستكشاف",
  "more": "المزيد",
  "language": "اللغة",
  "theme": "المظهر",
  "light": "فاتح",
  "dark": "داكن"
}