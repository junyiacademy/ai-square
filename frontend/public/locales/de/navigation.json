{
  "home": "Startseite",
  "relations": "Beziehungen",
  "ksa": "KSA-Rahmenwerk",
  "assessment": "KI-Bewertung",
  "pbl": "PBL-Lernen",
  "assessmentHistory": "Bewertungsverlauf",
  "history": "Verlauf",
  "signIn": "Anmelden",
  "signOut": "Abmelden",
  "userRole": {
    "student": "Sch√ºler",
    "teacher": "Lehrer",
    "admin": "Administrator"
  },
  "darkMode": "In den dunklen Modus wechseln",
  "lightMode": "In den hellen Modus wechseln",
  "toggleTheme": "Thema wechseln"
}