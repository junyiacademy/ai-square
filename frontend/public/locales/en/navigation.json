{
  "home": "Home",
  "relations": "Relations",
  "ksa": "KSA Framework",
  "assessment": "AI Assessment",
  "signIn": "Sign in",
  "signOut": "Sign out",
  "userRole": {
    "student": "Student",
    "teacher": "Teacher",
    "admin": "Administrator"
  },
  "darkMode": "Switch to dark mode",
  "lightMode": "Switch to light mode",
  "toggleTheme": "Toggle theme"
}