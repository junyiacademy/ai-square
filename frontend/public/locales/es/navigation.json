{
  "home": "Inicio",
  "relations": "Relations",
  "ksa": "KSA Framework",
  "signIn": "Iniciar sesión",
  "signOut": "Cerrar sesión",
  "userRole": {
    "student": "Estudiante",
    "teacher": "Profesor",
    "admin": "Administrador"
  }
}