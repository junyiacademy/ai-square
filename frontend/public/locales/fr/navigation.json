{
  "home": "Accueil",
  "relations": "Relations",
  "ksa": "Cadre KSA",
  "assessment": "Évaluation",
  "pbl": "Apprentissage PBL",
  "assessmentHistory": "Historique d'évaluation",
  "history": "Historique",
  "signIn": "Se connecter",
  "signOut": "Se déconnecter",
  "userRole": {
    "student": "Étudiant",
    "teacher": "Enseignant",
    "admin": "Administrateur"
  },
  "darkMode": "Passer en mode sombre",
  "lightMode": "Passer en mode clair",
  "toggleTheme": "Changer de thème",
  "historySubtitle": "Consultez tous vos enregistrements d'historique d'apprentissage",
  "totalActivities": "Total des activités",
  "filterAll": "Tous",
  "noHistory": "Aucun enregistrement d'apprentissage pour le moment",
  "startAssessment": "Commencer l'évaluation",
  "startPBL": "Commencer l'apprentissage PBL",
  "startNewPBL": "Commencer un nouvel apprentissage PBL",
  "footer": {
    "product": "Produit",
    "resources": "Ressources",
    "company": "Entreprise",
    "about": "À propos d'AI Square",
    "description": "AI Square est une plateforme d'apprentissage basée sur Git pour l'éducation à l'alphabétisation en IA, prenant en charge 9 langues dans le monde.",
    "journey": "Parcours utilisateur",
    "roadmap": "Feuille de route du produit",
    "feedback": "Commentaires",
    "rights": "Tous droits réservés.",
    "privacy": "Politique de confidentialité",
    "terms": "Conditions d'utilisation",
    "contact": "Contact",
    "languages": "Langues",
    "domains": "Domaines IA",
    "competencies": "Compétences",
    "madeWith": "Fait avec ❤️ pour l'éducation à l'alphabétisation en IA"
  }
}