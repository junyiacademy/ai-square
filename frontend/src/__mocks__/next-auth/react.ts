export const useSession = jest.fn(() => ({
  data: null,
  status: 'unauthenticated'
}));

export const signIn = jest.fn();
export const signOut = jest.fn();
export const getCsrfToken = jest.fn();
export const getSession = jest.fn();
export const getProviders = jest.fn();

export const SessionProvider = ({ children }: { children: React.ReactNode }) => children;