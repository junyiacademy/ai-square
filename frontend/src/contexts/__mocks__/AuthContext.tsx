import React from 'react';

// Mock AuthContext for testing
export const mockAuthValue = {
  user: null,
  loading: false,
  error: null,
  signIn: jest.fn(),
  signOut: jest.fn(),
  signUp: jest.fn(),
};

export const AuthContext = React.createContext(mockAuthValue);

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <AuthContext.Provider value={mockAuthValue}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = jest.fn(() => mockAuthValue);