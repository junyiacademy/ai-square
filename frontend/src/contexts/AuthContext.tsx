'use client';

import React, { createContext, useContext, useState, useEffect, useCallback } from 'react';
import { useRouter } from 'next/navigation';

interface User {
  id: number;
  email: string;
  role: string;
  name: string;
}

interface AuthContextType {
  user: User | null;
  isLoggedIn: boolean;
  isLoading: boolean;
  tokenExpiringSoon: boolean;
  login: (credentials: { email: string; password: string; rememberMe?: boolean }) => Promise<{ success: boolean; error?: string }>;
  logout: () => Promise<void>;
  checkAuth: () => Promise<void>;
  refreshToken: () => Promise<boolean>;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const router = useRouter();
  const [user, setUser] = useState<User | null>(null);
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  const [tokenExpiringSoon, setTokenExpiringSoon] = useState(false);

  const clearAuthState = useCallback(() => {
    localStorage.removeItem('isLoggedIn');
    localStorage.removeItem('user');
    localStorage.removeItem('ai_square_session');
    setUser(null);
    setIsLoggedIn(false);
    setTokenExpiringSoon(false);
    window.dispatchEvent(new CustomEvent('auth-changed'));
  }, []);

  const updateAuthState = useCallback((userData: User) => {
    setUser(userData);
    setIsLoggedIn(true);
    localStorage.setItem('isLoggedIn', 'true');
    localStorage.setItem('user', JSON.stringify(userData));
    window.dispatchEvent(new CustomEvent('auth-changed'));
  }, []);

  const checkAuth = useCallback(async () => {
    try {
      const response = await fetch('/api/auth/check', {
        credentials: 'include'
      });
      
      if (!response.ok) {
        throw new Error('Auth check failed');
      }
      
      const data = await response.json();
      
      if (data.authenticated && data.user) {
        updateAuthState(data.user);
        setTokenExpiringSoon(data.tokenExpiringSoon || false);
      } else {
        clearAuthState();
      }
    } catch (error) {
      console.error('Error checking auth:', error);
      
      // Fallback to localStorage only if API fails
      const storedAuth = localStorage.getItem('isLoggedIn');
      const storedUser = localStorage.getItem('user');
      
      if (storedAuth === 'true' && storedUser) {
        try {
          const parsedUser = JSON.parse(storedUser);
          setUser(parsedUser);
          setIsLoggedIn(true);
        } catch {
          clearAuthState();
        }
      } else {
        clearAuthState();
      }
    } finally {
      setIsLoading(false);
    }
  }, [clearAuthState, updateAuthState]);

  const login = useCallback(async (credentials: { email: string; password: string; rememberMe?: boolean }) => {
    try {
      const response = await fetch('/api/auth/login', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        credentials: 'include',
        body: JSON.stringify(credentials),
      });

      const data = await response.json();

      if (data.success && data.user) {
        updateAuthState(data.user);
        
        if (data.sessionToken) {
          localStorage.setItem('ai_square_session', data.sessionToken);
        }
        
        return { success: true };
      } else {
        return { success: false, error: data.error || 'Login failed' };
      }
    } catch (error) {
      console.error('Login error:', error);
      return { success: false, error: 'Network error' };
    }
  }, [updateAuthState]);

  const logout = useCallback(async () => {
    try {
      await fetch('/api/auth/logout', {
        method: 'POST',
        credentials: 'include'
      });
    } catch (error) {
      console.error('Logout error:', error);
    }
    
    clearAuthState();
    router.push('/login');
  }, [clearAuthState, router]);
  
  const refreshToken = useCallback(async (): Promise<boolean> => {
    try {
      const response = await fetch('/api/auth/refresh', {
        method: 'POST',
        credentials: 'include'
      });
      
      if (response.ok) {
        const data = await response.json();
        if (data.success) {
          await checkAuth();
          return true;
        }
      }
    } catch (error) {
      console.error('Token refresh error:', error);
    }
    
    return false;
  }, [checkAuth]);

  // 初始化時檢查登入狀態
  useEffect(() => {
    // 先從 localStorage 快速設置狀態，避免 UI 閃動
    const storedAuth = localStorage.getItem('isLoggedIn');
    const storedUser = localStorage.getItem('user');
    
    if (storedAuth === 'true' && storedUser) {
      try {
        const parsedUser = JSON.parse(storedUser);
        setUser(parsedUser);
        setIsLoggedIn(true);
      } catch (error) {
        console.error('Error parsing stored user:', error);
      }
    }
    
    // 然後進行 API 驗證
    checkAuth();
  }, [checkAuth]);

  // 監聽其他 tab 的登入狀態變化
  useEffect(() => {
    const handleStorageChange = (e: StorageEvent) => {
      if (e.key === 'isLoggedIn' || e.key === 'user') {
        const newLoggedInStatus = localStorage.getItem('isLoggedIn');
        const newUserData = localStorage.getItem('user');
        
        if (newLoggedInStatus === 'true' && newUserData) {
          try {
            const parsedUser = JSON.parse(newUserData);
            setUser(parsedUser);
            setIsLoggedIn(true);
          } catch {
            clearAuthState();
          }
        } else {
          clearAuthState();
        }
      }
    };

    const handleAuthChange = () => {
      // 當其他組件觸發 auth-changed 事件時，重新檢查狀態
      const storedAuth = localStorage.getItem('isLoggedIn');
      const storedUser = localStorage.getItem('user');
      
      if (storedAuth === 'true' && storedUser) {
        try {
          const parsedUser = JSON.parse(storedUser);
          setUser(parsedUser);
          setIsLoggedIn(true);
        } catch {
          clearAuthState();
        }
      } else {
        clearAuthState();
      }
    };

    window.addEventListener('storage', handleStorageChange);
    window.addEventListener('auth-changed', handleAuthChange);
    
    return () => {
      window.removeEventListener('storage', handleStorageChange);
      window.removeEventListener('auth-changed', handleAuthChange);
    };
  }, [clearAuthState]);

  // Token 快過期時自動刷新
  useEffect(() => {
    if (tokenExpiringSoon && isLoggedIn) {
      refreshToken();
    }
  }, [tokenExpiringSoon, isLoggedIn, refreshToken]);

  const value: AuthContextType = {
    user,
    isLoggedIn,
    isLoading,
    tokenExpiringSoon,
    login,
    logout,
    checkAuth,
    refreshToken,
  };

  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
}

export function useAuth(): AuthContextType {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}