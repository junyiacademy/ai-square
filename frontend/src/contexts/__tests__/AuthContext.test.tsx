/**
 * Tests for AuthContext - Auth State Synchronization
 */

import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { AuthProvider, useAuth } from '../AuthContext';
import { useRouter } from 'next/navigation';

// Mock dependencies
jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));

// Mock fetch
global.fetch = jest.fn();

// Test component that uses useAuth
function TestComponent() {
  const { user, isLoggedIn, login } = useAuth();
  
  return (
    <div>
      <div data-testid="login-status">{isLoggedIn ? 'Logged In' : 'Not Logged In'}</div>
      <div data-testid="user-email">{user?.email || 'No User'}</div>
      <button onClick={() => login({ email: 'test@example.com', password: 'test123' })}>
        Login
      </button>
    </div>
  );
}

describe('AuthContext', () => {
  const mockPush = jest.fn();
  
  beforeEach(() => {
    jest.clearAllMocks();
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
    localStorage.clear();
    
    // Mock successful auth check
    (global.fetch as jest.Mock).mockImplementation((url) => {
      if (url === '/api/auth/check') {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ 
            authenticated: false,
            user: null 
          }),
        });
      }
      if (url === '/api/auth/login') {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({
            success: true,
            user: {
              id: 1,
              email: 'test@example.com',
              role: 'student',
              name: 'Test User'
            },
            sessionToken: 'test-token'
          }),
        });
      }
      return Promise.reject(new Error('Not found'));
    });
  });

  it('should initialize with logged out state', async () => {
    render(
      <AuthProvider>
        <TestComponent />
      </AuthProvider>
    );
    
    await waitFor(() => {
      expect(screen.getByTestId('login-status')).toHaveTextContent('Not Logged In');
      expect(screen.getByTestId('user-email')).toHaveTextContent('No User');
    });
  });

  it('should update state immediately after successful login', async () => {
    render(
      <AuthProvider>
        <TestComponent />
      </AuthProvider>
    );
    
    // Initial state
    expect(screen.getByTestId('login-status')).toHaveTextContent('Not Logged In');
    
    // Click login button
    const loginButton = screen.getByText('Login');
    
    await act(async () => {
      loginButton.click();
    });
    
    // After login, state should be updated immediately
    await waitFor(() => {
      expect(screen.getByTestId('login-status')).toHaveTextContent('Logged In');
      expect(screen.getByTestId('user-email')).toHaveTextContent('test@example.com');
    });
    
    // Check localStorage was updated
    expect(localStorage.getItem('isLoggedIn')).toBe('true');
    expect(localStorage.getItem('user')).toBeTruthy();
    const storedUser = JSON.parse(localStorage.getItem('user') || '{}');
    expect(storedUser.email).toBe('test@example.com');
  });

  it('should trigger auth-changed event after login', async () => {
    const authChangedHandler = jest.fn();
    window.addEventListener('auth-changed', authChangedHandler);
    
    render(
      <AuthProvider>
        <TestComponent />
      </AuthProvider>
    );
    
    const loginButton = screen.getByText('Login');
    
    await act(async () => {
      loginButton.click();
    });
    
    await waitFor(() => {
      expect(authChangedHandler).toHaveBeenCalled();
    });
    
    window.removeEventListener('auth-changed', authChangedHandler);
  });

  it('should persist login state from localStorage on mount', async () => {
    // Set localStorage before rendering
    localStorage.setItem('isLoggedIn', 'true');
    localStorage.setItem('user', JSON.stringify({
      id: 1,
      email: 'cached@example.com',
      role: 'student',
      name: 'Cached User'
    }));
    
    render(
      <AuthProvider>
        <TestComponent />
      </AuthProvider>
    );
    
    // Should immediately show logged in state from localStorage
    expect(screen.getByTestId('login-status')).toHaveTextContent('Logged In');
    expect(screen.getByTestId('user-email')).toHaveTextContent('cached@example.com');
  });

  it('should handle auth check API validation', async () => {
    // Mock auth check to return authenticated
    (global.fetch as jest.Mock).mockImplementation((url) => {
      if (url === '/api/auth/check') {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ 
            authenticated: true,
            user: {
              id: 1,
              email: 'api@example.com',
              role: 'student',
              name: 'API User'
            }
          }),
        });
      }
      return Promise.reject(new Error('Not found'));
    });
    
    render(
      <AuthProvider>
        <TestComponent />
      </AuthProvider>
    );
    
    // Should update from API check
    await waitFor(() => {
      expect(screen.getByTestId('login-status')).toHaveTextContent('Logged In');
      expect(screen.getByTestId('user-email')).toHaveTextContent('api@example.com');
    });
  });
});