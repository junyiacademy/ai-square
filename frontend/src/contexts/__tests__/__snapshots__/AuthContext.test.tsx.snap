// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AuthContext should match snapshot 1`] = `
<div>
  Test
</div>
`;
